### ISSUE TITLE
tabcmd command hangs without login info in powershell

### ISSUE THREAD

**Issue by kulicny** [tableau/tabcmd#201]
I am testing tabcmd 2.0 prior to implementing in a production environment and have run into an issue. In my current 1.0 scripts, I first connect to a session and then subsequent commands can be issued without providing credentials. While testing 2.0, I can successfully connect using login but issuing any further commands results in that command just hanging. There is nothing in the log file and using debug mode also results in no useful information. If I issue the second command and again supply the login info, it works as expected and I can export my view so I'm guessing, based on the behavior, that it is waiting for credentials to be supplied. I have tried adding the -cookie command to the login to see if that would help, but it did not. I'm wondering if this is by design and that the session is no longer maintained, and credentials must be sent for each export? Or is there something else needed in order to get subsequent commands to maintain the connection until logout is specifically requested?

**Comment by jacalata** [comment]
That is not how it's supposed to work, that's strange. Can you share the export command you run?

**Comment by jacalata** [comment]
could you run tabcmd -v and share the version info?

**Comment by kulicny** [comment]
Tableau Server Command Line Utility v2.0.7

Ken Ulicny

**Comment by jacalata** [comment]
I just re-read this and noticed that you are running it within powershell, is that right? I'll have to check that out. 

If you run the same command from windows cmd, do you see the same behavior?

**Comment by kulicny** [comment]
Yes, from powershell.  I can try running from windows command for a test but we are using powershell and I have so many scripts, I won’t be able to conver them nor do I want to do so. I’ll send you the results next week.

Ken Ulicny

On Dec 9, 2022, at 8:59 PM, Jac ***@***.***> wrote:

﻿

CAUTION: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe.

I just re-read this and noticed that you are running it within powershell, is that right? I'll have to check that out.

If you run the same command from windows cmd, do you see the same behavior?

**Comment by r0b3r7** [comment]
Hi All - I have some issue from powershell on all 2.0.x versions. Anyone have some ideas how to solve it?

**Comment by jacalata** [comment]
Investigating this — the root cause is that after `tabcmd login`, the saved session in `~/tableau_auth.json` stores the auth token but subsequent commands don't reuse it to skip re-authentication. Instead they fall into a code path that calls `getpass.getpass()`, which blocks indefinitely in non-interactive environments like PowerShell scripts.

**Workaround until this is fixed:** Use a Personal Access Token (PAT) instead of `--password` for your login step. PAT-based auth does not call `getpass`, so subsequent commands work correctly in scripts:

```powershell
tabcmd login --server https://your-server --username you@example.com --token-name MyTokenName --token-value your-token-value
tabcmd export ...  # works without re-supplying credentials
```

You can generate a PAT in Tableau Server under Account Settings → Personal Access Tokens.
